### PR TITLE
Configure RPC endpoints specific to transports 

### LINKED ISSUE #1037: Configure RPC endpoints specific to transports
### Describe the feature

Currently, the `RpcModuleBuilder` only has one `build` function

https://github.com/paradigmxyz/reth/blob/c23a2215b0ddb7660c45b2f32bd422491f03c3c9/crates/net/rpc-builder/src/lib.rs#L92-L95

However, we need to configure endpoints for a specific means of transport, for example `--http.api eth,debug --ws-.api eth`

This is configured via:

https://github.com/paradigmxyz/reth/blob/c23a2215b0ddb7660c45b2f32bd422491f03c3c9/crates/net/rpc-builder/src/lib.rs#L125

Which will be configured via `ServerBuilder`

https://github.com/paradigmxyz/reth/blob/c23a2215b0ddb7660c45b2f32bd422491f03c3c9/crates/net/rpc-builder/src/lib.rs#L166

So the ideal workflow would be `ServerBuilder -> RpcModuleBuilder` where `ServerBuilder` knows what to enable per transport.

`RpcModuleBuilder::build` needs to be refactored that it can return an `RpcModule` per transport if enabled

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by literallymarvellous**
resolves: #1037 

Draft.

Not sure how to use ```RethRpcModule``` and ```RpcModuleConfig``` in ```RpcModuleBuilder::build``` yet. Im guessing each enum variants it's own methods?.

### CODE REVIEW

File: crates/net/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/net/rpc-builder/src/lib.rs:95]: this goes in the right direction but is yet exactly what we need.

what we want here, instead of `Vec<RpcServer>` is a type that has `Option<RpcModuleConfig>` per transport. 
something like

```rust
struct TransportRpcModuleConfig {
  ws: Option<RpcModuleConfig>,
  HTTP, ipc
}
```

then build should return a similar type
```rust
struct TransportRpcModules {
  ws: Option<RpcModule>
 ....
} 
```

File: crates/net/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/net/rpc-builder/src/lib.rs:220]: I'd prefer types that have a field per server instead

see comment above.

**Review by mattsse — CHANGES_REQUESTED** [review]
this will require some incremental changes because we're not entirely sure how it will look like at this point.

Really appreciate your effort here.

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [inline crates/net/rpc-builder/src/lib.rs:225]: I decided to remove the ```RpcServerEnum``` and change ```RpcServer``` this to 
```rust
pub struct RpcServer {
    /// http server
    http: Option<Server>,
    /// ws server
    ws: Option<Server>,
    /// ipc server
    ipc: Option<IpcServer>,
}
```
so the ```build``` function on ```RpcServerBuilder``` returns ```Result<RpcServer, RpcError>``` not ```Result<Vec<RpcServer>, RpcError>```

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [inline crates/net/rpc-builder/src/lib.rs:250]: Same here, to 
```rust
pub struct RpcServerHandle {
    http: Option<ServerHandle>,
    ws: Option<ServerHandle>,
    ipc: Option<ServerHandle>,
}
```

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [reply crates/net/rpc-builder/src/lib.rs:95]: For this, we'll have to parse something like ```--http.api eth,debug --ws-.api eth``` into a ```(http, [eth, debug]), (ws, [eth])``` type struct, that constructs ```TransportRpcModuleConfig```.

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [reply crates/net/rpc-builder/src/lib.rs:225]: Also change
 ```rust     
pub async fn start(self, methods: impl Into<Methods>) -> Result<RpcServerHandle, RpcError>
```
to 
 ```rust     
pub async fn start(self, modules: TransportRpcModules) -> Result<RpcServerHandle, RpcError> 
```

File: crates/net/rpc-builder/src/lib.rs
Comment (gakonst) [inline crates/net/rpc-builder/src/lib.rs:191]: These should be unwrap or a default value. For http/ws 8545, for ipc should be in the global datadir? https://github.com/paradigmxyz/reth/blob/acac82fcb3bf31c2e70fe1aba0e9eca6c7ad021f/bin/reth/src/dirs.rs#L13-L22

Comment (gakonst) [inline crates/net/rpc-builder/src/lib.rs:191]:
```suggestion
        let socket_addr = self.http_ws_addr.unwrap();
        let ipc_path = self.ipc_server_path.unwrap();
```

File: crates/net/rpc-builder/src/lib.rs
Comment (gakonst) [inline crates/net/rpc-builder/src/lib.rs:195]: Let's have separate configs for HTTP and for WS, sometimes one may not want to bring up both, or may want to expose different ports for each one.

**Review by gakonst — CHANGES_REQUESTED** [review]
good start- i rebased the PR to get rid of the large merge commits

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [reply crates/net/rpc-builder/src/lib.rs:191]: Not sure, so I'll leave ti as unwrap() for now.

File: crates/net/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/net/rpc-builder/src/lib.rs:100]: can leave as todo! for now
```suggestion
```

File: crates/net/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/net/rpc-builder/src/lib.rs:259]: this won't work. because there's no Into<Methods> type that is also Copy.

I'm ok with leaving this entire function as a `todo!` because I think we need to figure out all the config settings first.

**Review by mattsse — CHANGES_REQUESTED** [review]
it'd be great if we can keep this PR limited to the new Config types, to unblock further work.
need to figure out how to construct everything and what settings we'll need first.

File: crates/net/rpc-builder/src/lib.rs
Comment (literallymarvellous) [reply crates/net/rpc-builder/src/lib.rs:259]: alright. will make changes.

**Review by mattsse — CHANGES_REQUESTED** [review]
looks like something went wrong while rebasing this PR?

PR changes to rpc look good, but this PR needs to be only these changes.

**Review by mattsse — APPROVED** [review]

**Comment by literallymarvellous** [comment]
@mattsse Hey, I'm sorry, I think I've messed up the rebase.

**Comment by mattsse** [comment]
uh something went terrible wrong here.
I realised we moved the rpc crate which probably lead to this mess. sorry about that.

I rebased it rather sloppy as well but should do, thank you for this!

**Review by mattsse — APPROVED** [review]

**Comment by literallymarvellous** [comment]
Alright, thanks for handling it.

**Comment by gakonst** [comment]
I think this is a bit light? Would add some more code to it so that it does something before merging, I thought this would actually spin up things perhaps

**Pull request merged**
